### PR TITLE
CORS & CSP fixups

### PR DESCRIPTION
**Pull request by 73rhodes**
Introduces a number of CSP & CORS -related fixes:
* catch CSP error in `web_socket` to allow a fallback
* Fix CORS isUsable to return only truly successful connections
* Stop calling isUsable for disallowed connection types
* Fix CORS headers handling

### CODE REVIEW

File: src/transport/web_socket.js [hunk 1]
```diff
@@ -53,7 +53,13 @@ var WebSocket = assign(Class(Transport, {
...
     this._state = this.CONNECTING;
 
-    var socket = this._createSocket();
+    var socket;
+
+    try {
+      socket = this._createSocket();
+    } catch (e) {
+      // catch CSP error to allow transport to fallback to next connType
+    }
```
Comment (73rhodes) [inline src/transport/web_socket.js]: Catches CSP error in web_socket to allow fallback.

File: src/transport/transport.js [hunk 2]
```diff
@@ -177,10 +177,8 @@ var Transport = assign(Class({ className: 'Transport',
       if (array.indexOf(disabled, connType) >= 0)
         return resume();
 
-      if (array.indexOf(allowed, connType) < 0) {
-        klass.isUsable(dispatcher, connEndpoint, function() {});
+      if (array.indexOf(allowed, connType) < 0)
```
Comment (73rhodes) [inline src/transport/transport.js:180]: Stop calling isUsable for disallowed connection types.

File: src/transport/cors.js [hunk 3]
```diff
@@ -77,8 +77,9 @@ var CORS = assign(Class(Transport, {
     if (global.XMLHttpRequest) {
       var xhr = new XMLHttpRequest();
       return callback.call(context, xhr.withCredentials !== undefined);
+    } else {
+      return callback.call(context, false);
```
Comment (73rhodes) [inline src/transport/cors.js]: stylistic preference / lint-related use of if/else

**Comment by jcoglan** [comment]
There are several distinct changes in this PR and I'd appreciate some more context on what the motivation behind each of them is. Especially if any of them are a response to issues encountered in production, I would like to understand the nature of said issues.

- The changes to `spec/javascript/node_adapter_spec.js` look like a refactoring to inline the `handleResponse()` function, but don't actually change the tests' behaviour. The PR also does not change the actual behaviour of the server with regard to responding to `OPTIONS` requests. So I'm not sure what the purpose of this change is.
- The change to `src/transport/cors.js` looks it does not change any behaviour, it just moves the tail of a function into the `else` of the preceding `if` statement.
- The change to `src/transport/web_socket.js` looks like it should have no effect, because `_createSocket()` already wraps the creation of the socket in a `try`/`catch` that's intended to handle any CSP problems. I don't see what further wrapping the call to this function in an exception handler achieves.

The change to `src/transport/transport.js` is the only one that changes any behaviour, but it breaks some intended functionality. The initial handshake can only be done via `long-polling` or `callback-polling` -- that's what the `allowed` parameter represents. However, while the handshake is taking place, we want to start the work to figure out if other transports are usable, so that there is less time to wait after the handshake completes and we want to switch to a more capable transport.

Is this code causing a problem in production, and if so what is the nature of that problem?

**Comment by 73rhodes** [comment]
Thanks for looking over the draft!

> There are several distinct changes in this PR and I'd appreciate some more context on what the motivation behind each of them is. Especially if any of them are a response to issues encountered in production, I would like to understand the nature of said issues.
> 
> * The changes to `spec/javascript/node_adapter_spec.js` look like a refactoring to inline the `handleResponse()` function, but don't actually change the tests' behaviour. The PR also does not change the actual behaviour of the server with regard to responding to `OPTIONS` requests. So I'm not sure what the purpose of this change is.

It looks like this was based on a [comment thread](https://github.com/faye/faye/pull/483#issuecomment-329286396) where we used one approach to solve an issue but it was then resolved in the main upstream repository, which probably makes our change redundant.

> * The change to `src/transport/cors.js` looks it does not change any behaviour, it just moves the tail of a function into the `else` of the preceding `if` statement.

Yep, agreed. It looks like this was an incidental change of the previously mentioned CORS headers fix we applied. I can certainly drop this from the PR.

> * The change to `src/transport/web_socket.js` looks like it should have no effect, because `_createSocket()` already wraps the creation of the socket in a `try`/`catch` that's intended to handle any CSP problems. I don't see what further wrapping the call to this function in an exception handler achieves.

Indeed. It looks like when we originally committed this to our fork, there was no try/catch on the `_createSocket()` call yet. I can remove this from the PR as well.

> The change to `src/transport/transport.js` is the only one that changes any behaviour, but it breaks some intended functionality. The initial handshake can only be done via `long-polling` or `callback-polling` -- that's what the `allowed` parameter represents. However, while the handshake is taking place, we want to start the work to figure out if other transports are usable, so that there is less time to wait after the handshake completes and we want to switch to a more capable transport.
> 
> Is this code causing a problem in production, and if so what is the nature of that problem?

I reached out to the author of this change @jugarrit for more context! We should have an update shortly 🤞

**Comment by 73rhodes** [comment]
Unfortunately, we no longer have the context on why this change was made, but it looks like the only remaining modification in this PR was essentially meant as an optimization, albeit a misguided one that removed intended functionality according to your comment @jcoglan ... Closing this PR!

**Pull request closed without merging**